[REVIEW]Distributed TF-IDF transformer

VibhuJawa opened the pull request:
This pr closes https://github.com/rapidsai/cuml/issues/2598 . 

After this PR, we should be able to use this in distributed TF-IDF Pipe-Lines.  (HashingVectorizer+TF-IDF Transformer)

See Gist for working examples. 

- [Distributed example of Doc frequency search on Tweet dataset](https://gist.github.com/VibhuJawa/32e74a2ed8f595bab3b19a289905dcb8) which got almost the same results as the old single-GPU tfidf [implementation](https://gist.github.com/VibhuJawa/e8cefb11cafcd9f176891465988d7f71). 
- [Accuracy on news dataset as sklearn](https://gist.github.com/VibhuJawa/fd777785b7e38dc312b6556c42ec1a00)

Todo:
- [x] Verify the same results on a small example
- [x] Add tests
- [x] Benchmark
- (`330 ms` on 4 GPU vs `369 ms` on a single GPU) on a 4Million Tweet dataset
  - Needs bigger datasets to see impact. 
  - Will evaluate https://github.com/rapidsai/tpcx-bb/tree/master/tpcx_bb/queries/q28 once PR lands. 
- [x] Code Clean up (Especially around cumlArray)  


CC: @randerzander

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

VibhuJawa commented on python/cuml/feature_extraction/_tfidf.py:
```diff
@@ -90,11 +102,60 @@ class TfidfTransformer:
+        # TODO Finding it difficult to
+        # map below to cumlarray
+        # removed trailing underscores for now
+        self.df = df
```
This is a temporary array that we create.  What is the best way to mark this as a member of this class  ?

dantegd replied:
what do you mean exactly with marking it as member of the class?

VibhuJawa replied:
So I want to make these a `super private` attribute (https://github.com/rapidsai/cuml/issues/2456#issuecomment-666106406)  but I run into a bunch of `CumlArray` related test failures.  

What is the best way to create a temporary array that I can just keep as a cupy array or is there a better convention i should follow ?

VibhuJawa replied:
Currently, I am getting a failure in tests here. 

https://github.com/rapidsai/cuml/blob/196cde0b36834849393281e54d4b9823a461a0d1/python/cuml/feature_extraction/_tfidf_vectorizer.py#L202

```python
              pass
        elif (supported_type is not None):
            # Is this an estimated property? If so, should always be CumlArray
>           assert not name.endswith("_"), "Invalid Estimated Array-Like \
                Attribute! Estimated attributes should always be CumlArray. \
                Attribute: '{}' In: {}".format(name, self.__repr__())
E           AssertionError: Invalid Estimated Array-Like                 Attribute! Estimated attributes should always be CumlArray.                 Attribute: 'idf_' In: TfidfTransformer()
E           assert not True
E            +  where True = <built-in method endswith of str object at 0x7feaab54d6f0>('_')
E            +    where <built-in method endswith of str object at 0x7feaab54d6f0> = 'idf_'.endswith

conftest.py:117: AssertionError
========================================================================= short test summary info ==========================================================================
FAILED test_text_feature_extraction.py::test_tfidf_vectorizer_idf_setter - AssertionError: Invalid Estimated Array-Like                 Attribute! Estimated attributes s...
============================================================================
```


I am not really sure how are we supposed to store and create members if we want to derive from the `Base` class ? 

Like, in this case, the setter creates a sparse diagonal array while the getter should return a cupy array and there is not really interoperability of input and output are sparse matrices.


Sorry for all the questions, kind of confused by this .

dantegd replied:
Ok the problem is that this is a very different attribute than the majority of attributes we're checking that are cumlarray (and doesn't return the sparse array), so the check doesn't apply right now. And overall tfidf and quite a few classes in feature_extraction are quite different than the bulk of cuML, so the checks don't apply as they are. The tests (and cumlarray management) will change/improve in 0.16, so this type of scenario will be taken into account. For now I would suggest adding an exception here

https://github.com/rapidsai/cuml/blob/196cde0b36834849393281e54d4b9823a461a0d1/python/cuml/test/conftest.py#L120

essentially  adding before that line something like 

```python
if name != idf_:
```
so that this parameter is not checked at least for now. I think afterwards with the improvements we can add a flag or parameter for classes that inherit from `Base` but want more flexibility in `CumlArray` usage.

VibhuJawa replied:
> that inherit from Base but want more flexibility in CumlArray usage

This will be great.   

The other question I have is what is the recommended way to store temporary super-private objects when inheriting from the base ? 

Is the recommendation to just store them as cuml_array and access the underlying cupy array using `to_output(out_type)` . ?

dantegd replied:
it depends, but in general it is perfectly fine to store it as a pure cupy array for this case as far as I can see.

VibhuJawa replied:
Resolved the idf issue by below: https://github.com/rapidsai/cuml/pull/2698/commits/85ac30c673444b3c8c80c0e78089aa0199413d65

Have started a more detailed discussion about this at https://github.com/rapidsai/cuml/issues/2704#issuecomment-675216038 . 

Let's continue the discussion there.

VibhuJawa commented:
rerun tests.

VibhuJawa commented:
This is now ready for a first review. CC: @cjnolet / @dantegd

cjnolet commented on python/cuml/feature_extraction/_tfidf.py:
```diff
@@ -30,7 +32,15 @@ def _sparse_document_frequency(X):
+    """
+        Returns the valid dtype for tf-idf transoformer
+    """
+    dtype = X.dtype if X.dtype in FLOAT_DTYPES else cp.float32
```
Can you copy in the `FLOAT_DTYPES` directly from sklearn?

Aside from tests and benchmarks, we have been trying to keep sklearn dependencies out of the Python.

VibhuJawa replied:
Resolved by https://github.com/rapidsai/cuml/pull/2698/commits/e6033f7e06e0930a6b497dd4fa8b326e41fc33b0

cjnolet commented on python/cuml/feature_extraction/_tfidf.py:
```diff
@@ -91,11 +101,53 @@ class TfidfTransformer:
+        n_features
+        df(document frequency)
+        """
+        # Should not have a cost if allready sparse
```
Small typo
```suggestion
        # Should not have a cost if already sparse
```

cjnolet commented on python/cuml/test/conftest.py:
```diff
@@ -110,7 +110,13 @@ def patched__setattr__(self, name, value):
 
-        if (supported_type == CumlArray):
+        if name == 'idf_':
+            # We skip this test because idf_' fot tfidf setter returns
```
Small typo
```suggestion
            # We skip this test because idf_' for tfidf setter returns
```

cjnolet commented on python/cuml/feature_extraction/_tfidf.py:
```diff
@@ -30,7 +32,15 @@ def _sparse_document_frequency(X):
-class TfidfTransformer:
+def _get_dtype(X):
+    """
+        Returns the valid dtype for tf-idf transoformer
```
Small typo
```suggestion
        Returns the valid dtype for tf-idf transformer
```

cjnolet commented on python/cuml/dask/feature_extraction/text/tfidf_transformer.py:
```diff
@@ -0,0 +1,208 @@
+        self._set_internal_model(s_TfidfTransformer(**kwargs))
+
+    @staticmethod
+    @with_cupy_rmm
```
I had tested this at one point and recall decorators not getting propagated to the workers. My testing was a Dask versions ago. Have you tested this recently just to make sure? Just to be safe, we can use the context manager inside the function to guarantee it's used.

VibhuJawa replied:
So I just verified it and it seems to be working (see [gist](https://gist.github.com/VibhuJawa/451795f57489dba67a41c6f6700a7908)) 


But we can still skip it as we anyways decorate the underlying function too. 
https://github.com/rapidsai/cuml/pull/2698/files#diff-27c6ae1bf7c794b7e79b86dbddc0f6baR113-R131 . 

Happy to keep/remove it, let me know .

cjnolet replied:
Awesome! If it's working then I'm okay keeping it as it is.

VibhuJawa commented:
@cjnolet , Have addressed all reviews apart from this [one](https://github.com/rapidsai/cuml/pull/2698#discussion_r478621957) as it seems to be working. 


Will update it if needed based on directions. Should be good for a final pass I think now.

VibhuJawa commented:
Rerun Tests.

VibhuJawa commented:
@cjnolet,  This should be good to go now. :-)
